source/wrangling.md

```diff
@@ -1152,1262 +960,890 @@ glue("census_popn", "{0:,.0f}".format(35151728))
-# In the code above,
-# the first line is just the `tidy_lang` data frame with a pipe.
-# The pipe passes the left-hand side (`tidy_lang`) to the first argument of the function on the right (`filter`),
-# so in the `filter` function you only see the second argument (and beyond).
-# Then again after `filter` there is a pipe, which passes the result of the `filter` step
-# to the first argument of the `select` function.
+```{table} Basic summary statistics
+:name: tab:basic-summary-statistics
+| Function | Description |
+| -------- | ----------- |
+| `count` | The number of observations (rows) |
+| `mean` | The mean of the observations |
+| `median` | The median value of the observations |
+| `std` | The standard deviation of the observations |
+| `max` | The largest value in a column |
+| `min` | The smallest value in a column |
+| `sum` | The sum of all observations |
+| `agg` | Aggregate multiple statistics together |
+| `describe` | a summary |
 ```
 
-As you can see, both of these approaches&mdash;with and without chaining&mdash;give us the same output, but the second
-approach is clearer and more readable.
-
 +++
-
-### Chaining more than two functions
-
 +++
 
-Chaining can be used with any method in Python. 
-Additionally, we can chain together more than two functions. 
-For example, we can chain together three functions to: 
 
-- extract rows (`df[]`) to include only those where the counts of the language most spoken at home are greater than 10,000, 
-- extract only the columns (`.loc`) corresponding to `region`, `language` and `most_at_home`, and
-- sort the data frame rows in order (`.sort_values`) by counts of the language most spoken at home 
-from smallest to largest.
+> **Note:** In `pandas` DataFrame, the value `NaN` is often used to denote missing data.
+> By default, when `pandas` calculates summary statistics (e.g., `max`, `min`, `sum`, etc),
+> it ignores these values. If you look at the documentation for these functions, you will
+> see an input variable `skipna`, which by default is set to `skipna=True`. This means that
+> `pandas` will skip `NaN` values when computing statistics.
+
+### Calculating summary statistics on dataframes
 
-```{index} pandas.DataFrame; sort_values
+What if you want to calculate summary statistics on an entire dataframe? Well,
+it turns out that the functions in {numref}`tab:basic-summary-statistics`
+can be applied to a whole data frame!
+For example, we can ask for the number of rows that each column has using `count`.
+```{code-cell} ipython3
+region_lang.count()
+```
+Not surprisingly, they are all the same. We could also ask for the `mean`, but
+some of the columns in `region_lang` contain string data with words like `"Vancouver"`
+and `"Halifax"`---for these columns there is no way for `pandas` to compute the mean.
+So we provide the keyword `numeric_only=True` so that it only computes the mean of columns with numeric values. This
```

Comment:
How annoying that pandas is starting to require this, I quite liked the convenient of not having to specify it... it seems like a reasonable default to only take the average of numerical columns... I wonder what their rationale is.